design/decisions/0004-entity-dregister-api.md

```diff
@@ -0,0 +1,244 @@
+# Entity Deregister API
+
+* Date: __2024-03-14__
+* Status: __New__
+
+## Background
+
+All the entities (devices and services) registered with %%te%% are stored with the MQTT broker as retained messages,
+with their metadata spread across multiple topics.
+For example, for a child device `child01`, the registration message is stored on the `te/device/child01//` topic,
+its twin data is stored across several `te/device/child01///twin/<twin-data-type>` topics,
+its command metadata stored across several `te/device/child01///cmd/<cmd-type>` topics and so on.
+
+On top of that, when a device has services associated with it, or has nested child devices,
+they all have their own respective registration topics and respective metadata topics.
+So, deregistering a device involves deregistering itself, its metadata and
+the complete entity hierarchy that is associated with it.
+
+## Challenges
+
+* Metadata and data associated with a given entity spread across multiple topics, must be cleared together.
+* When services and child devices are linked to a device, deleting that entity involves deleting all linked entities,
+  to avoid leaving orphan entities behind.
+* The parent and child devices may not share a common topic prefix, making it difficult to do it purely from an MQTT level.
+  For example, a device `te/device/child01//` may have an associated service `te/device/child01/service/service01`
+  and nested child devices `te/device/child11//` and `te/device/child12//`.
+  Even though the parent device and service has a common prefix `te/device/child01/#`, there is no such common prefix
+  between the parent device and its child devices.
+  And when custom topic schemes are used, even the device and its service may not have any common prefix available.
+
+## Finalized Solution Proposal
+
+Entities can be deregistered by clearing their registration message from the broker, publishing empty retained messages to their topic id, as follows:
+
+```sh
+tedge mqtt pub -r te/device/child01// ''
+```
+
+This message will result in the deregistration of that target entity and all the nested entities (services and child devices) linked to it.
+
+Different `tedge` components will react to the deregistration messages as follows:
```

Comment:
So, if I understand correctly:

- The user emits a single MQTT message to prune a hierarchy of entities.
- This clearing message is an empty retained message published on the root topic of the pruned hierarchy
- The mappers propagate entity removal to the cloud, pruning all attached entities (services, child devices, twin metadata, capabilities, ...)
- The agent has to responsibility to clear the local view of the entities. It cascades clearing messages along the pruned hierarchy ensuring these sub-entities will not be re-created on mapper restart.

This is okay, but we need to be aware that combined with auto-registration this might lead to strange hierarchy if the device is restarted before the agent gets a chance to prune all the sub-entities. Some will be re-created by the mapper but using the auto-registration logic.